include/tvm/internal/RangeCounting.h

```diff
@@ -0,0 +1,121 @@
+ * represented by {(1,+), (5,-)} where the lower limit (opening, denoted by +)
+ * is included and the upper limit (closing, denoted by -) is excluded.
+ * A list of such limits is kept as a representation of the current state.
+ * The number of appearances of an integer is encoded by the number of + and -
+ * appearing to reach that number from the start of this list. For example, if
+ * the list is {(1,+), (3,+), (5,-), (6,+), (7,-), (7,-), (9,+), (10,-)}, all
+ * number up to 0 don't appear, 1 and appear once, 3 and 4 appear twice, 5
+ * appears once, 6 appears twice, 7 and 8 don't appear, 9 appears once and all
+ * numbers from 10 don't appear. The resulting range representation is
+ * { Range(1,7), Range(9,1) }.
+ */
+class TVM_DLLAPI RangeCounting
+{
+public:
+  /** The lower (included) or upper (excluded) limit of a range*/
+  struct Limit
+  {
+    enum Type
+    {
+      Lower = -1,
+      Cut = 0,
+      Upper = 1
+    };
+
+    Limit(int i, Type type) : i_(i), type_(type) {}
+
+    int i_;     // Value of the limit
+    Type type_; // The type of the limit
+
+    bool operator==(const Limit & other) const { return (i_ == other.i_ && type_ == other.type_); }
+    bool operator<(const Limit & other) const { return i_ < other.i_ || (i_ == other.i_ && type_ < other.type_); }
+    bool operator<=(const Limit & other) const { return i_ < other.i_ || (i_ == other.i_ && type_ <= other.type_); }
+    bool operator>(const Limit & other) const { return i_ > other.i_ || (i_ == other.i_ && type_ > other.type_); }
+    bool operator>=(const Limit & other) const { return i_ > other.i_ || (i_ == other.i_ && type_ >= other.type_); }
+
+    friend std::ostream & operator<<(std::ostream & os, const Limit & lim)
+    {
+      os << "(" << lim.i_ << ", " << ((lim.type_ == Lower) ? "+" : ((lim.type_ == Cut) ? "|" : "-")) << ")";
+      return os;
+    }
+  };
+
+  /** Add a range to the counting. Return true if this changes the outuput of
+   * \c ranges(false) (it always changes the output of \c range(true) ).
+   */
+  bool add(const Range & r);
+  /** Remove a range from the counting. Return true if this changes the outuput
+   * of \c ranges() (it always changes the output of \c range(true) ).
+   */
+  bool remove(const Range & r);
+
+  bool empty() const { return limits_.size() == 0; }
+
+  /** Get a representation of number appearing as a list of ranges.
+   *
+   * \param splitOnDepthDiff If true, ranges are split on count differences and
+   * on Limit::Cut (i.e. two ranges touching but not overlapping will be returned
+   * separatly, not meerged).
+   */
+  const std::vector<Range> & ranges(bool splitOncountDiff = false) const;
+  /** Get the underlying representation as a list of limits.*/
+  const std::list<Limit> & limits() const;
+
+private:
+  using It = std::list<Limit>::iterator;
+
+  /** Forward to first element in limits_ that would come after \p val.
+   * Returns \c true if depth went to 0 in the process
```

Comment:
`s/0/\p depthCut/`

A more helpful documentation would be to say that it returns `true` if the "compact" representation of the range needs to be updated?

Reply:
I fixed the doc in 32599c3 for `0`

You're right that this is the use made of the return type, but with respect to the function itself, it's only about encountering the prescribed depth.